tasks/Task.cpp

```diff
@@ -2,48 +2,413 @@
+        [&](rtc::PeerConnection::GatheringState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::GatheringState::Complete:
+                mState.peer_connection.gathering_state = Complete;
+            case rtc::PeerConnection::GatheringState::InProgress:
+                mState.peer_connection.gathering_state = InProgress;
+            case rtc::PeerConnection::GatheringState::New:
+                mState.peer_connection.gathering_state = NewGathering;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onSignalingStateChange(
+        [&](rtc::PeerConnection::SignalingState state)
+        {
+            switch (state)
+            {
+            case rtc::PeerConnection::SignalingState::HaveLocalOffer:
+                mState.peer_connection.signaling_state = HaveLocalOffer;
+            case rtc::PeerConnection::SignalingState::HaveLocalPranswer:
+                mState.peer_connection.signaling_state = HaveLocalPranswer;
+            case rtc::PeerConnection::SignalingState::HaveRemoteOffer:
+                mState.peer_connection.signaling_state = HaveRemoteOffer;
+            case rtc::PeerConnection::SignalingState::HaveRemotePranswer:
+                mState.peer_connection.signaling_state = HaveRemotePranswer;
+            case rtc::PeerConnection::SignalingState::Stable:
+                mState.peer_connection.signaling_state = Stable;
+            default:
+                break;
+            }
+        });
+
+    peer_connection->onLocalDescription(
+        [&, remote_peer_id](rtc::Description description)
+        {
+            Json::Value message;
+            message["protocol"] = "one-to-one";
+            message["to"] = remote_peer_id;
+            message["action"] = description.typeString();
+            message["data"]["from"] = _local_peer_id.get();
+            message["data"]["description"] = string(description);
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+            {
+                Json::FastWriter fast;
+                ws->send(fast.write(message));
+            }
+        });
+
+    peer_connection->onLocalCandidate(
+        [&, remote_peer_id](rtc::Candidate candidate)
+        {
+            Json::Value message;
+            message["protocol"] = "one-to-one";
+            message["to"] = remote_peer_id;
+            message["action"] = "candidate";
+            message["data"]["from"] = _local_peer_id.get();
+            message["data"]["candidate"] = string(candidate);
+            message["data"]["mid"] = candidate.mid();
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+            {
+                Json::FastWriter fast;
+                ws->send(fast.write(message));
+            }
+        });
+
+    return peer_connection;
 }
 
-Task::~Task()
+void Task::configurePeerDataChannel(string const &remote_peer_id)
 {
+    mPeerConnection->onDataChannel(
```

Comment:
Still wrong. The issue is that `onDataChannel` is called only on the "passive" peer. The active peer (the one that creates the data channel) will not get the onDataChannel event

However, in the 'active peer' case, you have to set up the callbacks on the data channel object itself.